### PR TITLE
created function to support printing from Discord

### PR DESCRIPTION
**Pull request by phuvu0107**
We added a new command on the discord bot so this PR will support on it.

### CODE REVIEW

**Review by evanugarte — COMMENTED** [review]
merge dev into the branch

File: printer/downloadFile.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ */
+async function downloadFileFromURL(url) {
+  await axios({
+    url: url,
+    method: 'GET',
+    responseType: 'stream'
+  })
+    .then(response => {
+      const fileName = 
+        Math.random().toString(36).substring(2, 15) + Math.random().toString(23).substring(2, 5);
```
Comment (evanugarte) [inline printer/downloadFile.js]: this is crazy can we do anything better than this

File: printer/downloadFile.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      const filePath = `/tmp/${fileName}.pdf`;
+      fs.writeFileSync(filePath, '');
+      response.data.pipe(fs.createWriteStream(filePath));
+      return filePath;
+    })
+    .catch(error => {
+      console.log(error);
+    });
+}
```
Comment (evanugarte) [inline printer/downloadFile.js]: add logger.info and logger.error here, if it worked log where the file was downloaded to, if it didnt work log the error

File: printer/downloadFile.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+const axios = require('axios');
+const fs = require('fs');
+
+/**
+ * Download PDF file from URL.
+ * @param {Sring} url : PDF file url
+ * @returns filePath: the path to downloaded PDF file
+ */
+async function downloadFileFromURL(url) {
```
Comment (evanugarte) [inline printer/downloadFile.js]: shouldnt this be returning a promise? we call this function then await either the path to the file or false

File: printer/downloadFile.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      method: 'GET',
+      responseType: 'stream'
+    })
+      .then(response => {
+        const fileName = uuidv4();
+        const filePath = `/tmp/${fileName}.pdf`;
+        fs.writeFileSync(filePath, '');
+        const fileStream = fs.createWriteStream(filePath);
+        response.data.pipe(fileStream);
+        console.log('Downloaded file location: ', filePath);
```
Comment (evanugarte) [inline printer/downloadFile.js]: use logger here, not console.log, see [`tools/writeToSQS.js:1`](https://github.com/SCE-Development/Quasar/blob/831de7a7f3b07d23f2bb651992ee1c930d1e13a1/tools/writeToSQS.js#L1) as an example

File: printer/downloadFile.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+        resolve({success: true, pdfFilePath: filePath});
+      })
+      .catch(error => {
+        console.log(error);
+        resolve({ success: false });
```
Comment (evanugarte) [inline printer/downloadFile.js]:
```suggestion
        resolve(filePath);
      })
      .catch(error => {
        console.log(error);
        resolve(false);
```

File: printer/PrintHandler.js [hunk 6]
```diff
@@ -57,39 +58,46 @@ function determinePrinterForJob() {
...
+      deleteFile(fileNo);
+      sqs.deleteMessage(deleteParams, (err) => {
+        if (err) throw err;
       });
-  });
+    }); 
+  } else {
+    const fileDownloaded = await downloadFileFromURL(data.Body.fileURL);
+    !fileDownloaded.success && console.log('Double check the URL');
+    filePath = fileDownloaded.pdfFilePath;
```
Comment (evanugarte) [inline printer/PrintHandler.js]: if it wasnt a success should we continue with the below printing code?

**Review by neelp03 — APPROVED** [review]

**Review by ShaivPandya — APPROVED** [review]

File: printer/PrintHandler.js [hunk 7]
```diff
@@ -57,39 +58,46 @@ function determinePrinterForJob() {
...
-  const { fileNo, copies, pageRanges } = data.Body;
   const pages = pageRanges === 'NA' ? '' : '-P ' + pageRanges;
-  const path = `/tmp/${fileNo}.pdf`;
+  let filePath = `/tmp/${fileNo}.pdf`;
 
-  const paramers = {
-    Bucket: PRINTING_BUCKET_NAME,
-    Key: `folder/${fileNo}.pdf`,
-  };
+  if (!fromDiscord) {
```
Comment (akshtshrma24) [inline printer/PrintHandler.js]: Add a comment here, explaining what this does/why you did it, short comment

File: printer/PrintHandler.js [hunk 8]
```diff
@@ -57,39 +58,46 @@ function determinePrinterForJob() {
...
+        ReceiptHandle: data.ReceiptHandle,
+      };
+      deleteFile(fileNo);
+      sqs.deleteMessage(deleteParams, (err) => {
+        if (err) throw err;
       });
-  });
+    }); 
+  } else {
+    const fileDownloaded = await downloadFileFromURL(data.Body.fileURL);
```
Comment (akshtshrma24) [inline printer/PrintHandler.js:147]: add one here as well, a reader who doesnt know much about the file would be lost aka mee

File: printer/downloadFile.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        resolve(filePath);
+      })
+      .catch(error => {
+        logger.error('Download Unsuccessfully! ', error);
+        resolve(false);
+      });
+  });
+}
+
+module.exports = downloadFileFromURL;
```
Comment (evanugarte) [inline printer/downloadFile.js]: if u wanna import it like
```js
const { downloadFileFromURL } = require('./downloadFile');
```

then we must do
```suggestion
module.exports = { downloadFileFromURL };
```

File: printer/downloadFile.js [hunk 10]
```diff
@@ -0,0 +1,35 @@
+const axios = require('axios');
+const fs = require('fs');
+const { v4: uuidv4 } = require('uuid');
+const logger = require('../util/logger.js');
+/**
```
Comment (evanugarte) [inline printer/downloadFile.js]:
```suggestion


/**
```

File: printer/PrintHandler.js [hunk 11]
```diff
@@ -57,39 +58,51 @@ function determinePrinterForJob() {
...
+    });
+  // If the file is from Discord, download the file using URL.
+  } else {
+    const fileDownloaded = await downloadFileFromURL(data.Body.fileURL);
+    if(!fileDownloaded) return;
+    filePath = fileDownloaded.pdfFilePath;
+  }
+  
+  // Print downloaded file using command then delete that file.
+  exec(
```
Comment (evanugarte) [inline printer/PrintHandler.js:151]: the delete message from sqs function should be moved to below this exec

File: printer/PrintHandler.js [hunk 12]
```diff
@@ -27,6 +28,38 @@ const params = {
...
+ */
+async function downloadFileFromURL(url) {
+  return new Promise( ( resolve ) => {
+    axios({
+      url: url,
+      method: 'GET',
+      responseType: 'stream'
+    })
+      .then(response => {
+        const attachmentId = url.split('/')[5];
```
Comment (evanugarte) [inline printer/PrintHandler.js:50]:
```suggestion
        // discord attachment URLs look like below, where id and name
        // are unique to the uploaded attachment
        // https://cdn.discordapp.com/attachments/<server>/<id>/<name>
        const attachmentId = url.split('/')[5];
```

**Review by evanugarte — APPROVED** [review]

**Pull request merged**